src/ucp/core/ucp_mm.c

```diff
@@ -1305,3 +1412,349 @@ ucs_status_t ucp_mem_reg_md_map_update(ucp_context_h context)
+
+            iter = kh_put(ucp_context_imported_mem_rcaches_hash,
+                          context->imported_mem_rcaches, remote_uuid, &ret);
+            ucs_assertv((ret != UCS_KH_PUT_FAILED) &&
+                        (ret != UCS_KH_PUT_KEY_PRESENT), "ret %d", ret);
+
+            kh_value(context->imported_mem_rcaches, iter) = rcache;
+        } else {
+            rcache = existing_rcache;
+        }
+
+        status = ucp_memh_rcache_get(rcache, address, length, mem_type,
+                                     UCP_MEM_FLAG_IMPORTED, &memh);
+        if (status != UCS_OK) {
+            goto err_rcache_destroy;
+        }
+
+        user_memh->parent = memh;
+    } else {
+        memh   = user_memh;
+        rcache = NULL;
+    }
+
+    memh->reg_id      = reg_id;
+    memh->remote_uuid = remote_uuid;
+    status            = ucp_memh_import_attach(context, memh,
+                                               attach_params_array,
+                                               attach_params_num);
+    if (status != UCS_OK) {
+        goto err_memh_free;
+    }
+
+out:
+    UCP_THREAD_CS_EXIT(&context->mt_lock);
+    return status;
+
+err_memh_free:
+    if (rcache != NULL) {
+        ucs_rcache_region_put_unsafe(rcache, &memh->super);
+    }
+err_rcache_destroy:
+    if ((rcache != NULL) && (rcache != existing_rcache)) { 
+        /* Rcache was allocated here - remove it from hash and destroy */
+        iter = kh_get(ucp_context_imported_mem_rcaches_hash,
+                      context->imported_mem_rcaches, remote_uuid);
+        ucs_assert(iter != kh_end(context->imported_mem_rcaches));
+        kh_del(ucp_context_imported_mem_rcaches_hash,
+               context->imported_mem_rcaches, iter);
+        ucs_rcache_destroy(rcache);
+    }
+    goto out;
+}
+
+static ucs_status_t
+ucp_memh_import(ucp_context_h context, const void *export_mkey_buffer,
+                ucp_mem_h *memh_p)
+{
+    ucs_rcache_t *rcache = NULL;
+    const void *p        = export_mkey_buffer;
+    const void *end;
+    ucp_md_index_t md_index;
+    ucp_mem_h memh, rcache_memh;
+    ucp_md_index_t remote_md_index;
+    ucs_status_t status;
+    ucp_md_map_t remote_md_map, local_md_map;
+    ucp_memh_import_attach_params_t *attach_params_array;
+    unsigned attach_params_num;
+    ucs_memory_type_t mem_type;
+    uint16_t flags;
+    uint16_t UCS_V_UNUSED mem_info_parsed_size;
+    uint64_t remote_uuid;
+    const void *tl_mkey_buf;
+    ucs_rcache_region_t *rregion;
+    khiter_t iter;
+    void *address;
+    size_t length;
+    uint16_t memh_info_size;
+    uint64_t reg_id;
+
+    ucs_assert(export_mkey_buffer != NULL);
+
+    /* Common memory handle information */
+    memh_info_size = ucp_memh_info_size_unpack(&p);
+    end            = UCS_PTR_BYTE_OFFSET(export_mkey_buffer, memh_info_size);
+    ucs_assert(memh_info_size != 0);
+
+    flags         = *ucs_serialize_next(&p, uint16_t);
+    remote_md_map = *ucs_serialize_next(&p, ucp_md_map_t);
```

Comment:
we need to use strict datatypes for packing/unpacking, otherwise wire compat will break if `ucp_md_map_t` changes

Reply:
I saw that `ucp_md_map_t` is used in other places

Reply:
fixed